### PR TITLE
An importable modulesPath, once again

### LINKED ISSUE #8188: Bring back modulesPath
I'm thinking of bringing this line back:

https://github.com/NixOS/nixpkgs/commit/f69ce50529c5108d4dababfd221652ace55264bd#diff-e29cca50467118cb72b2d2c8f97cd7caL60

and getting rid of nixos/modules/misc/extra-arguments.nix. Does anyone have objections? I fully buy the ability to pass arguments in the module system using `_module.args`, but if it doesn't work (and breaks in a horrible way) for the most common use case (`modulesPath` in imports), that seems bad and the cost of restoring the old way seems pretty low.

Anyone have objections? I can make the change if nobody objects. Ideally, I'd like it in 15.06, since it's getting in the way of some work I'm trying to squeeze into 15.06.

cc @nbp @shlevy @edolstra

### PR DESCRIPTION
**Pull request by copumpkin**
This is the simplest way I can think of to make it happen and work the way it used to (which doesn't seem that bad). Alternative to #8189 and should resolve #8188.

This feels like an extremely noisy problem that has a dozen valid solutions. @edolstra @shlevy @nbp probably care the most. I'm building my functionality on top of this but will swap out something else if people prefer it. I mostly just want to put this problem behind us and move on :smile:

### CODE REVIEW

**Comment by copumpkin** [comment]
Also related to #7835.

**Comment by shlevy** [comment]
:+1: avoids the "special treatment of NixOS in the module system" problem and doesn't do scope magic. Bikeshedding possible on the name of `specialArgs` but I don't really care personally.

File: lib/modules.nix [hunk 1]
```diff
@@ -17,6 +17,10 @@ rec {
      evalModules) and the less declarative the module set is. */
   evalModules = { modules
                 , prefix ? []
+                , # This should only be used for special arguments that need to be evaluated
+                  # when resolving module structure (like in imports). For everything else,
+                  # use _module.args (and Mastercard).
```
Comment (domenkozar) [inline lib/modules.nix]: No need for product placements in good jokes :P

Comment (copumpkin) [inline lib/modules.nix]: Damn, I was hoping nobody would notice! You just cost me $5

**Comment by copumpkin** [comment]
Also, after merging this, I'd probably want to go back and change the various `<nixpkgs/...>` imports back to `modulesPath`. I've run into lots of confusing issues due to those `NIX_PATH` imports floating around in nixpkgs.

**Comment by edolstra** [comment]
I don't think we should change those imports back to `modulesPaths`. Keeping `modulesPath` around for backwards compatibility is one thing, but it shouldn't be used for new deployments.

**Comment by edolstra** [comment]
:+1: on merging btw.

**Comment by copumpkin** [comment]
Okay, just waiting on @nbp then.

@edolstra I still think there's a legitimate use case for this thing, and that's the one I'm encountering with my EC2 image: if I want to both build an image/VM test from a config (in my current tree) and copy that config into the resulting image (will be in /etc/nixos) for maximum realism/avoiding DRY, I don't think I have another option. Am I missing a good way to make that work?

**Comment by nbp** [comment]
Couldn't you use the `-I` command line argument?  I am not sure I understand your problem with the `<nixpkgs/…>` notation.

**Comment by copumpkin** [comment]
Sure I can, but:
1) it's a trap; no indication ahead of time or even after the fact that a
particular call to nix-build is going to need -I. The system will just
quietly do the wrong thing unless you notice
2) why make people jump through hoops for this, what is so bad about
keeping the old behavior that actually served a purpose?
3) the tests could be running in hydra. I don't get to specify -I there and
don't have guarantees about where it's pointing

**Comment by copumpkin** [comment]
In short, `<nixpkgs>`-style paths are basically an untracked impurity that can deeply and subtly change your expression's behavior, allowing what would otherwise be a pure expression (anywhere you import it) to access global state. I'm fine with them around "entry points", but if you ever find yourself processing a `<nixpkgs>` behind an `import`, I think you're in for some trouble/pain. The implementation of `eval-config` probably imports `configuration.nix` right now, but `configuration.nix` is still the "entry point" as far as the user is concerned. That's where my #8190 proposal is coming from, if anyone missed that.

**Comment by edolstra** [comment]
I don't understand the problem. How is writing `<nixpkgs/...>` jumping through hoops compared to the totally ad hoc `modulesPath`? And how will it do the wrong thing?

Regarding Hydra, it's completely guaranteed what `<nixpkgs>` points to, namely, the `nixpkgs` input of the jobset.

**Comment by copumpkin** [comment]
The problem is just that I got bitten repeatedly when trying to run what seemed like a perfectly innocent expression in #8013. I'd type `nix-build -A tests.ec2 release.nix` and think it was doing the right thing, but deep within my import expression someone imported `<nixpkgs/...>` and was calling my system `nixpkgs` rather than the directory I was running in.

**Comment by copumpkin** [comment]
How about `scopedImport` has an option to provide additional (or modify existing) `NIX_PATH` components in the evaluated expression? Then when I import something I can be sure that its `nixpkgs` is pointing to the right place?

**Comment by copumpkin** [comment]
@nbp anyway, can we split the discussion into two parts?

1. whether this is a "good" (or simple, at least) way to get `modulesPath` back
2. whether people should want `modulesPath` going forward

Both @shlevy and @edolstra seem to be okay with 1. I still have some work to do to convince Eelco at least of 2, but that can happen later (or I can be convinced that it isn't useful), but I'd rather not hold up 1 for that. Do you have a view on that?

**Comment by shlevy** [comment]
@copumpkin Let's just merge this for now.

**Comment by copumpkin** [comment]
Merging because 2/3 "module VIPs" have :+1:'d it :smile:. We can revert if it ends up being terribad.

@edolstra, @nbp, at some point (if I ever find some time to be on IRC while you guys are awake) we should talk about whether I'm just missing something in my desire to use this parameter.

**Pull request merged**